File: experiments/numpy/svd.py
import numpy as np
import scipy as sp
import time

np.show_config()

def measure_svd_time(h,w,svd_func,t=100):
    sum = 0
    for i in range(t):
        A = np.random.rand(h, w)
        start = time.time()
        U, S, Vt = svd_func(A)
        end = time.time()
        sum += end - start
    return sum / t


def compare_svd(h,w,t=100):
    print(f"{h} x {w} matrix SVD comparison over {t} trials:")
    svd_t=measure_svd_time(h,w, lambda A: sp.linalg.svd(A, full_matrices=False, compute_uv=True, lapack_driver='gesvd'),t=t)
    print(f"SVD Thin took {svd_t:.4f} seconds")
    svd_f=measure_svd_time(h,w, lambda A: sp.linalg.svd(A, full_matrices=True, compute_uv=True, lapack_driver='gesvd'),t=t)
    print(f"SVD Full took {svd_f:.4f} seconds")
    sdd_t=measure_svd_time(h,w, lambda A: sp.linalg.svd(A, full_matrices=False, compute_uv=True, lapack_driver='gesdd'),t=t)
    print(f"SVDDC Thin took {sdd_t:.4f} seconds")
    sdd_f=measure_svd_time(h,w, lambda A: sp.linalg.svd(A, full_matrices=True, compute_uv=True, lapack_driver='gesdd'),t=t)
    print(f"SVDDC Full took {sdd_f:.4f} seconds")
    print("")

compare_svd(400, 400)
compare_svd(200, 3000)

File: experiments/ndarray/src/bin/svd.rs
use ndarray::{Array1, Array2};
use ndarray_linalg::{JobSvd, SVD, SVDDC, TruncatedOrder, TruncatedSvd, random_using};
use num_traits::ToPrimitive;
use rand::{SeedableRng, rngs::SmallRng};
use std::{
    hint::black_box,
    time::{Duration, Instant},
};

extern crate blas_src as _;

fn measure_svd_time(
    h: usize,
    w: usize,
    trials: usize,
    svd_func: impl Fn(Array2<f64>) -> (Option<Array2<f64>>, Array1<f64>, Option<Array2<f64>>),
) -> f64 {
    let mut total_time = Duration::ZERO;

    let mut rng = SmallRng::seed_from_u64(0);

    for _ in 0..trials {
        // Create random matrix
        let x: Array2<f64> = random_using([h, w], &mut rng);

        let pre = Instant::now();

        black_box(svd_func(x));

        let post = Instant::now();

        total_time += post.duration_since(pre);
    }

    total_time.as_secs_f64() / trials.to_f64().unwrap()
}

fn compare_svd(h: usize, w: usize, trials: usize) {
    println!(
        "{} x {} matrix SVD comparison over {} trials:",
        h, w, trials
    );

    let svd_full = measure_svd_time(h, w, trials, |x| x.svd(true, true).unwrap());
    println!("SVD Full took {:.4} seconds", svd_full);

    let sdd_thin = measure_svd_time(h, w, trials, |x| x.svddc(JobSvd::Some).unwrap());
    println!("SDD Thin took {:.4} seconds", sdd_thin);

    let sdd_full = measure_svd_time(h, w, trials, |x| x.svddc(JobSvd::All).unwrap());
    println!("SDD Full took {:.4} seconds", sdd_full);

    let tsvd = measure_svd_time(h, w, trials, |x| {
        let (u, d, v) = TruncatedSvd::new(x, TruncatedOrder::Largest)
            .decompose(5)
            .unwrap()
            .values_vectors();
        (Some(u), d, Some(v))
    });
    println!("Truncated SVD took {:.4} seconds", tsvd);

    println!();
}

fn main() {
    compare_svd(400, 400, 100);
    compare_svd(200, 3000, 100); // 100 might be too long for large matrices
}
